keps/sig-node/753-sidecar-containers/README.md

```diff
@@ -843,24 +843,53 @@ examples:
 
 ### Termination of containers
 
-In alpha sidecar containers will be terminated as regular containers. No special
+In Alpha sidecar containers will be terminated as regular containers. No special
 or additional signals will be supported.
 
-We will collect feedback from alpha implementation and plan to improve
-termination in Beta. When Pods with sidecars are terminated:
+In Beta we have thought about introducing additional termination grace period fields
+to manage termination duration
+([draft proposal](https://docs.google.com/document/d/1B01EdgWJAfkT3l6CIwNwoskQ71eyuwet3mjiQrMQbU8))
+and leverage these fields to add reverse order termination of sidecar containers
+after the primary containers terminate.
 
+However, we decided on an alternative that doesn't require additional fields or hooks while keeping
+the desired behaviors when Pods with sidecars are terminated. While original approach works better
+with truly graceful termination where consistency is more important than time taken, proposed approach
+works for that scenario as well as a more and more popular scenario of limited time to terminate when
+graceful termination is set by external requirement and Pods needs to do best to gracefully terminate
+as much as possible (think of a Spot Instances with 30 seconds notification).
+
+Here is the proposed approach:
+1. Sidecar containers that have a `PreStop` hook will be notified when the Pod has begun terminating
+   by executing the `PreStop` hook. This happens at the same time as regular containers, and begins
+   the Pod's termination grace period countdown.
+2. Sidecar containers enter the `Terminated` state and are no longer restarted if they fail.
+3. Once the last primary container terminates, the last started sidecar container is notified by
+   sending a `SIGTERM` signal, even if `PreStop` hook hasn't completed yet.
+4. The next sidecar (in reverse order) is notified by sending a `SIGTERM` signal after the previous
```

Comment:
How long does each sidecar get between SIGTERM and SIGKILL ?  Or are you saying that it is ALL bounded by the pod's main TGPS ?  Considering that we (I) have previously written that serving apps SHOULD NOT self-exit, this means that many will receive SIGTERM and just wait for SIGKILL, leaving nothing for the sidecars.

Reply:
this is fine of they will not leave anything to sidecars. With the proposal sidecars will still receive the notification to clear buffers in a form of `preStop` hook. Once preStop received they can switch to small buffer mode AND start flushing all buffers.

The alternative approach that we discussed in alpha and linked as a google doc, if the main container simply ignores SIGTERM, sidecars are missing any notifications. So often logs will be missing and termination will not be as graceful as it could have been.